plugins/sinks/shield/sink.go

```diff
@@ -0,0 +1,180 @@
+type Config struct {
+	Host    string            `mapstructure:"host" validate:"required"`
+	Headers map[string]string `mapstructure:"headers"`
+}
+
+var info = plugins.Info{
+	Description: "Send user information to shield grpc service",
+	Summary:     summary,
+	Tags:        []string{"grpc", "sink"},
+	SampleConfig: heredoc.Doc(`
+	# The hostname of the shield service
+	host: shield.com:5556
+	# Additional headers send to shield, multiple headers value are separated by a comma
+	headers:
+	  X-Shield-Email: meteor@odpf.io
+      X-Other-Header: value1, value2
+	`),
+}
+
+type Sink struct {
+	plugins.BasePlugin
+	client Client
+	config Config
+	logger log.Logger
+}
+
+func New(c Client, logger log.Logger) plugins.Syncer {
+	s := &Sink{
+		logger: logger,
+		client: c,
+	}
+	s.BasePlugin = plugins.NewBasePlugin(info, &s.config)
+
+	return s
+}
+
+func (s *Sink) Init(ctx context.Context, config plugins.Config) error {
+	if err := s.BasePlugin.Init(ctx, config); err != nil {
+		return err
+	}
+
+	if err := s.client.Connect(ctx, s.config.Host); err != nil {
+		return fmt.Errorf("error connecting to host: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Sink) Sink(ctx context.Context, batch []models.Record) error {
+	for _, record := range batch {
+		asset := record.Data()
+		s.logger.Info("sinking record to shield", "record", asset.GetUrn())
+
+		userRequestBody, err := s.buildUserRequestBody(asset)
+		if err != nil {
+			return errors.Wrap(err, "failed to build shield payload")
+		}
+
+		if err = s.send(ctx, userRequestBody); err != nil {
+			return errors.Wrap(err, "error sending data")
+		}
+
+		s.logger.Info("successfully sinked record to shield", "record", asset.Name)
+	}
+
+	return nil
+}
+
+func (s *Sink) Close() (err error) {
+	return
+	//TODO: Connection closes even when some records are unpiblished
+	//TODO: return s.client.Close()
+}
+
+func (s *Sink) send(ctx context.Context, userRequestBody *sh.UserRequestBody) error {
+	for hdrKey, hdrVal := range s.config.Headers {
+		hdrVals := strings.Split(hdrVal, ",")
+		for _, val := range hdrVals {
+			val = strings.TrimSpace(val)
+			md := metadata.New(map[string]string{hdrKey: val})
+			ctx = metadata.NewOutgoingContext(ctx, md)
+		}
+	}
+
+	_, err := s.client.UpdateUser(ctx, &sh.UpdateUserRequest{
+		Id:   userRequestBody.Email,
+		Body: userRequestBody,
+	})
+	if err == nil {
+		return nil
+	}
+
+	if e, ok := status.FromError(err); ok {
+		err = fmt.Errorf("shield returns code %d: %v", e.Code(), e.Message())
+		switch e.Code() {
+		case codes.Unavailable:
+			return plugins.NewRetryError(err)
+		default:
+			return err
+		}
+	} else {
+		err = fmt.Errorf("not able to parse error returned %v", err)
+	}
+
+	return err
+}
+
+func (s *Sink) buildUserRequestBody(asset *assetsv1beta2.Asset) (*sh.UserRequestBody, error) {
+	data := asset.GetData()
+
+	var user assetsv1beta2.User
+	err := data.UnmarshalTo(&user)
+	if err != nil {
+		return &sh.UserRequestBody{}, errors.Wrap(err, "not a User struct")
+	}
```

Comment:
Can we move it to another method? basically if this line fails, we can just skip the record instead of failing the sink or the whole batch.

Reply:
Basically the goal is, if `Asset.Data` is not User then skip the whole process, instead of error.

Reply:
Understood, let me put a 'continue` in the loop of the Sink method, [line 82](https://github.com/odpf/meteor/blob/a97ad7997936c512ad18e20f073bc654d2ffd01a/plugins/sinks/shield/sink.go#L82) and [line 86](https://github.com/odpf/meteor/blob/a97ad7997936c512ad18e20f073bc654d2ffd01a/plugins/sinks/shield/sink.go#L86). This shall skip the users which have failure building the request or sending.

Reply:
Not on line 86, as we still need to send retry errors in case of send failure.